### PR TITLE
Configuration cleanup

### PR DESCRIPTION
**Pull request by ghost**
Configuration docs are rather messy now. [Deshts forum post](https://forums.spongepowered.org/t/object-serialization-what-am-i-missing/9708) showed some ways in which they were lacking and I'm pretty confident they can lose some fat in other places.

### CODE REVIEW

**Comment by Spongy** [comment]
A preview for this pull request is available at https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/aad53ee4/index.html.

Here are some links to the pages that were modified:

- modified https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/aad53ee4/plugin/configuration/loaders.html
- modified https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/aad53ee4/plugin/configuration/nodes.html
- modified https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/aad53ee4/plugin/configuration/paths.html
- modified https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/aad53ee4/plugin/configuration/serialization.html
- modified https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/aad53ee4/plugin/configuration/values.html

_Since the preview frequently changes, please link to [this comment](https://github.com/SpongePowered/SpongeDocs/pull/369#issuecomment-154859895), not to the direct url to the preview._

File: source/plugin/configuration/paths.rst [hunk 1]
```diff
@@ -12,7 +12,7 @@ character, leading to the final child node that holds a value, whether that valu
 another node. As an example, ``config.blockCheats.enabled`` is a path, leading to the node ``enabled``, under
 ``blockCheats``, which is under ``config``. In a configuration file, this would appear like so:
 
-.. code-block:: json
+.. code-block:: hocon
```
Comment (Tzky) [inline source/plugin/configuration/paths.rst]: I'm not sure what went wrong, but this code-block looks broken.

**Comment by saladoc** [comment]
I deem this PR fit for final review. Nitpicking welcome.

File: source/plugin/configuration/index.rst [hunk 2]
```diff
@@ -79,35 +84,18 @@ If you are unsure of what to set the value of ``sharedRoot`` to, consider the fo
     @DefaultConfig(sharedRoot = true)
     private ConfigurationLoader<CommentedConfigurationNode> configManager;
 
+    @Inject
+    @ConfigDir(sharedRoot = false)
+    private File privateConfigDir;
```
Comment (Minecrell) [inline source/plugin/configuration/index.rst]: You should either mention both or only `Path` - the recommended way is using `Path` now.

File: source/plugin/configuration/loaders.rst [hunk 3]
```diff
@@ -14,6 +14,11 @@ First, let's grab a new ``HoconConfigurationLoader`` that points to our configur
 
 .. code-block:: java
 
+    import java.io.File;
+    import ninja.leaping.configurate.commented.CommentedConfigurationNode;
+    import ninja.leaping.configurate.hocon.HoconConfigurationLoader;
+    import ninja.leaping.configurate.loader.ConfigurationLoader;
+
     File potentialFile = new File("config.conf");
```
Comment (Minecrell) [inline source/plugin/configuration/loaders.rst]: Same here, should use preferably `Path` (together with `setPath` below)

File: source/plugin/configuration/nodes.rst [hunk 4]
```diff
@@ -2,47 +2,184 @@
...
+~~~~~~~~~~~~
+
+Basic value types like ``int``, ``double``, ``boolean`` or ``String`` each have their own convenience getter method
+which will return the value or a default if the node does not contain a value of that type. Let's check if the server
+administrator wants our plugin to enable its blockCheats module by checking the value at the
+``modules.blockCheats.enabled`` path.
+
+.. code-block:: java
+
+    final boolean shouldEnable = rootNode.getNode("modules", "blockCheats", "enabled").getBoolean();
```
Comment (Minecrell) [inline source/plugin/configuration/nodes.rst]: Sometimes you use final in the examples here, sometimes you don't. Not sure if it's really important for the example.

File: source/plugin/injection.rst [hunk 5]
```diff
@@ -12,19 +12,23 @@ Dependency injection allows plugins to designate a few API types that will be in
...
+
+DefaultConfig (annotation on File, Path or ConfigurationLoader)
   Used to inject the plugin's specific configuration file: ``<Plugin#id>.conf``
 
 EventManager
   Manages the registration of event handlers and the dispatching of events.
 
-File
+File or Path
   This must have an additional annotation specifying specific file!
```
Comment (Minecrell) [inline source/plugin/injection.rst]: > This must have an additional annotation specifying specific file!

You haven't changed this, although this sentence doesn't make sense for me right now ("specifying specific file")

File: source/plugin/configuration/index.rst (hunk 2 above)
Comment (Zidane) [inline source/plugin/configuration/index.rst]: Only ```Path``` should be mentioned.

File: source/plugin/configuration/nodes.rst (hunk 4 above)
Comment (saladoc) [inline source/plugin/configuration/nodes.rst]: Actually I'm a little conflicted about whether the examples should use final. On one hand, it's a somewhat better style. On the other hand, it adds bloat to the examples.

File: source/plugin/configuration/index.rst [hunk 6]
```diff
@@ -63,51 +62,45 @@ If you are unsure of what to set the value of ``sharedRoot`` to, consider the fo
 * If you plan on having multiple configuration files (complex plugins) in the future, set the value to ``false``.
 * If you plan on having a single configuration file (less-complex plugins), set the value to ``true``.
 
+You can also obtain a ``Path`` instance pointing to the config directory instead of a particular file. Just
+have it injected using the ``@ConfigDir`` annotation, either with ``sharedRoot`` set to ``false`` for a plugin specific
+directory or to ``true`` to get the shared configuration directory.
+
+.. note::
+
+    While it may be possible to get a ``File`` instead of a ``Path``, configurate (and Sponge) recommend using ``Path``.
```
Comment (RobertHerhold) [inline source/plugin/configuration/index.rst]: capital C in Configurate

File: source/plugin/configuration/index.rst [hunk 7]
```diff
@@ -63,51 +62,45 @@ If you are unsure of what to set the value of ``sharedRoot`` to, consider the fo
...
+    @Inject
+    @ConfigDir(sharedRoot = false)
+    private Path privateConfigDir;
+
 .. warning::
 
-    When your plugin is running for the first time, the returned pathname may refer to a configuration file that does
-    not yet exist.
+    When your plugin is running for the first time, returned pathnames for configuration files and directories may not
+    yet exist. If you delegate all reading / writing of files to configurate, you do not need to worry about inexistant
```
Comment (RobertHerhold) [inline source/plugin/configuration/index.rst]: capital C in Configurate

Comment (RobertHerhold) [inline source/plugin/configuration/index.rst]: 'inexistant' --> 'nonexistent'

File: source/plugin/configuration/index.rst [hunk 8]
```diff
@@ -63,51 +62,45 @@ If you are unsure of what to set the value of ``sharedRoot`` to, consider the fo
...
-            ConfigurationLoader<ConfigurationNode> yamlReader =
-              YAMLConfigurationLoader.builder().setFile(bukkitConfigFile).build();
-            ConfigurationNode bukkitConfig = yamlReader.load();
-            configLoader.save(bukkitConfig);
-            if (!bukkitConfigFile.renameTo(new File(configDir, "config.yml.bukkit"))) {
-                logger.warn(lf(_("Could not rename old Bukkit configuration file to old name")));
-            }
-        }
-    }
+    plugins is HOCON. Conversion from YAML to HOCON can be automated by using one a ``YamlConfigurationLoader`` to load
```
Comment (RobertHerhold) [inline source/plugin/configuration/index.rst]: take out 'one'

File: source/plugin/configuration/nodes.rst [hunk 9]
```diff
@@ -2,47 +2,184 @@
...
-.. code-block:: java
-
-    File potentialFile = new File("config.conf");
-    ConfigurationLoader<CommentedConfigurationNode> loader = HoconConfigurationLoader.builder().setFile(potentialFile).build();
-    ConfigurationNode rootNode;
-    try {
-        rootNode = loader.load();
-    } catch(IOException e) {
-        rootNode = null;
+In memory, the the configuration is represented using ``ConfigurationNode``\ s. A ``ConfigurationNode`` either holds a
```
Comment (RobertHerhold) [inline source/plugin/configuration/nodes.rst]: I think @Inscrutable noted this in IRC, but 'the the' needs fixed

File: source/plugin/configuration/nodes.rst [hunk 10]
```diff
@@ -2,47 +2,184 @@
...
+value (like a number, a string or a list) or has child nodes, thus allowing for tree-like configuration structure.
+When using a ``ConfigurationLoader`` to load or create a new configration, it will return you the **root node**. It is
```
Comment (RobertHerhold) [inline source/plugin/configuration/nodes.rst]: Take out 'you'

File: source/plugin/configuration/nodes.rst [hunk 11]
```diff
@@ -2,47 +2,184 @@
...
+* Any basic value (see above)
+* Any ``List`` or ``Map`` of serializable types
+* The types ``java.util.UUID``, ``java.net.URL``, ``java.net.URI`` and ``java.util.regex.Pattern``
+* Any type that has been made serializable as described on :doc:`the config serialization page <serialization>`
+
+
+Defaults
+~~~~~~~~
+
+Unlike the Sponge API, the configurate library does not use ``Optional`` for values that might not be present but null.
```
Comment (RobertHerhold) [inline source/plugin/configuration/nodes.rst]: Captial C in Configurate

File: source/plugin/configuration/nodes.rst [hunk 12]
```diff
@@ -2,47 +2,184 @@
...
+
+This call will return ``false`` if either the value ``false`` is saved in the config or the value is not present in the
+config. Since those two cases are indistinguishable we have no simple way of setting our variable to ``false`` only if
+that is the value specified on the config. Unless we specify ``true`` as the default value.
+
+.. code-block:: java
+
+    boolean shouldEnable = rootNode.getNode("modules", "blockCheats", "enabled").getBoolean(true);
+
+Similar like this you can specify defaults on any value you get from the config, thus avoiding ``null`` returns or
```
Comment (RobertHerhold) [inline source/plugin/configuration/nodes.rst]: "Similar like this" --> "Similarly,"

File: source/plugin/configuration/nodes.rst [hunk 13]
```diff
@@ -2,47 +2,184 @@
...
+.. code-block:: java
+
+    String greeting = rootNode.getNode("messages", "greeting").getString("FLARD be with you good man!");
+
+    UUID mayor = rootNode.getNode("towns", "aFLARDia", "mayor")
+                            .getValue(TypeToken.of(UUID.class), somePlayer.getUniqueId());
+
+Another useful application of those defaults is that they can be copied to your configuration if needed. Upon creation
+of your root configuration node, you can create your ``ConfigurationOptions`` with ``setShouldCopyDefaults(true)``. If
+you did this, whenever you provide a default value, configurate will first check if the value you're trying to get is
```
Comment (RobertHerhold) [inline source/plugin/configuration/nodes.rst]: capital C Configurate

File: source/plugin/configuration/serialization.rst [hunk 14]
```diff
@@ -47,7 +51,11 @@ and one to create an object from a given configuration node.
         }
     }
 
-This ``TypeSerializer`` then must be registered with configurate.
+This ``TypeSerializer`` then must be registered with configurate. This can be done either globally by registering to
```
Comment (RobertHerhold) [inline source/plugin/configuration/serialization.rst]: capital C Configurate

**Pull request merged**